optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/DefaultSingleConstraintVerification.java

```diff
@@ -35,8 +46,53 @@ protected DefaultSingleConstraintVerification(
         this.scoreDirectorFactory = scoreDirectorFactory;
     }
 
+    private void assertDistinctPlanningIds(SolutionDescriptor<Solution_> solutionDescriptor, Object... facts) {
+        if (facts.length < 2) {
+            return;
+        }
+        // Group all entities by their @PlanningEntity-annotated class.
+        Map<Class<?>, List<Object>> entitiesByClassMap = Arrays.stream(facts)
+                .filter(fact -> solutionDescriptor.hasEntityDescriptor(fact.getClass()))
+                .collect(Collectors.groupingBy(fact -> {
+                    EntityDescriptor<Solution_> entityDescriptor =
+                            solutionDescriptor.findEntityDescriptor(fact.getClass());
+                    return entityDescriptor.getEntityClass();
+                }, ConcurrentHashMap::new, Collectors.toList()));
+        // Go over all the top-level classes, find and report duplicate IDs.
+        entitiesByClassMap.forEach((clz, clzFacts) -> {
+            MemberAccessor planningIdAccessor =
+                    ConfigUtils.findPlanningIdMemberAccessor(clz, solutionDescriptor.getDomainAccessType());
+            if (planningIdAccessor == null) {
+                return;
+            }
+            // Group entities by the same ID.
+            Map<Optional<Long>, List<Object>> entitiesWithSameIdMap = clzFacts.stream()
+                    .collect(Collectors.groupingBy(fact -> {
+                        Long planningId = (Long) planningIdAccessor.executeGetter(fact);
+                        return Optional.ofNullable(planningId); // Return as reference to prevent null keys.
+                    }));
+            // Eliminate those matches where there are no duplicate IDs and find the first one.
+            Optional<Map.Entry<Optional<Long>, List<Object>>> firstDuplicateIdEntry = entitiesWithSameIdMap.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue().size() > 1) // More than 1 instance means there is a duplicate ID.
+                    .findFirst();
+            firstDuplicateIdEntry.ifPresent(entry -> {
```

Comment:
I chose to only report the first entity class with duplicate instances.
So when the user fixes this failure in tests, they may need to fix another for another entity class.

I do consider this worthy. Otherwise the exception message would become even more massive than it already is.